UT4MasterServer/Controllers/SessionController.cs

```diff
@@ -49,121 +48,123 @@ public class SessionController : JsonAPIController
+				//
+				//       we could still support it, since ut understands it cuz its old and we don't
+				//       really need multi-factor auth. it is after all the way that ut's login screen
+				//       works when you start the game without launcher (and without UT4UU).
 
-					if (request.Password == null)
-					{
-						return ErrorInvalidRequest("password");
-					}
-
-					account = await accountService.GetAccountAsync(request.Username, request.Password);
-					if (account != null)
-					{
-						session = await sessionService.CreateSessionAsync(account.ID, clientID, SessionCreationMethod.Password);
-					}
-					else
-					{
-						return Unauthorized(new ErrorResponse
-						{
-							ErrorMessage = "Invalid credentials. Check your username and password and try again",
-						});
-					}
-
-					break;
-				}
-			default:
+				// EPIC is returning this ErrorResponse after checking username and password... This is better place IMO
+				if (!allowPasswordGrant)
 				{
 					return BadRequest(new ErrorResponse
 					{
-						ErrorCode = "errors.com.epicgames.common.oauth.unsupported_grant_type",
-						ErrorMessage = $"Unsupported grant type: {request.GrantType}",
-						NumericErrorCode = 1016,
+						ErrorCode = "errors.com.epicgames.common.oauth.unauthorized_client",
+						ErrorMessage = $"Sorry your client is not allowed to use the grant type {request.GrantType}. Please download and use UT4UU",
+						NumericErrorCode = 1015,
 						OriginatingService = "com.epicgames.account.public",
 						Intent = "prod",
-						ErrorDescription = $"Unsupported grant type: {request.GrantType}",
-						Error = "unsupported_grant_type",
+						ErrorDescription = $"Sorry your client is not allowed to use the grant type {request.GrantType}. Please download and use UT4UU",
+						Error = "unauthorized_client",
 					});
 				}
+
+				if (request.Username == null)
+				{
+					return ErrorInvalidRequest("username");
+				}
+
+				if (request.Password == null)
+				{
+					return ErrorInvalidRequest("password");
+				}
+
+				account = await accountService.GetAccountAsync(request.Username, request.Password);
+				if (account != null)
+					session = await sessionService.CreateSessionAsync(account.ID, clientID, SessionCreationMethod.Password);
+
+				break;
+			}
+			default:
+			{
+				return BadRequest(new ErrorResponse
+				{
+					ErrorCode = "errors.com.epicgames.common.oauth.unsupported_grant_type",
+					ErrorMessage = $"Unsupported grant type: {request.GrantType}",
+					NumericErrorCode = 1016,
+					OriginatingService = "com.epicgames.account.public",
+					Intent = "prod",
+					ErrorDescription = $"Unsupported grant type: {request.GrantType}",
+					Error = "unsupported_grant_type",
+				});
+			}
 		}
 
-		if (session == null) // only here to prevent null warnings, should never happen
+		if (session == null)
 		{
-			var message = $"Empty session. grant_type: {request.GrantType}";
+			var message = $"Invalid credentials";
```

Comment:
Are you sure this is only for `password` grant_type?

Reply:
credentials is any info used to authenticate a user. this includes `exchange_code` for `grant_type=exchange_code`, `username` and `password` for `grant_type=password`, etc.